src/roiextractors/memmapextractors.py

```diff
@@ -2,109 +2,118 @@
-        )
-
-        if verbose:
-            for ch in range(self.imaging.get_num_channels()):
-                print(f"Saving channel {ch}")
-                for i in tqdm(range(self.imaging.get_num_frames())):
-                    plane = self.imaging.get_frames(i, channel=ch)
-                    self._video[ch, i] = plane
-        else:
-            for ch in range(self.imaging.get_num_channels()):
-                for i in range(self.imaging.get_num_frames()):
-                    plane = self.imaging.get_frames(i, channel=ch)
-                    self._video[ch, i] = plane
-
-    @property
-    def filename(self):
-        return str(self._tmp_file)
-
-    def move_to(self, save_path):
-        save_path = Path(save_path)
-        if save_path.suffix != ".dat" and save_path.suffix != ".bin":
-            save_path = save_path.with_suffix(".dat")
-        if not save_path.parent.is_dir():
-            os.makedirs(save_path.parent)
-        shutil.move(self._tmp_file, str(save_path))
-        self._tmp_file = str(save_path)
-        self._video = np.memmap(
-            save_path,
-            shape=(
-                self.imaging.get_num_channels(),
-                self.imaging.get_num_frames(),
-                self.imaging.get_image_size()[0],
-                self.imaging.get_image_size()[0],
-            ),
-            dtype=self.imaging.get_dtype(),
-            mode="r",
-        )
-
-    @check_get_frames_args
-    def get_frames(self, frame_idxs, channel=0):
-        return self._video[channel, frame_idxs]
+        """Class for reading binary data.
+
+
+        Parameters
+        ----------
+        file_path : PathType
+            the file_path where the data resides.
+        frame_shape : tuple
+            The frame shape of the image determines how each frame looks. Examples:
+            (n_channels, rows, columns), (rows, columns, n_channels), (n_channels, columns, rows), etc.
+            Note that n_channels is 1 for grayscale and 3 for color images.
+        dtype : DtypeType
+            The type of the data to be loaded (int, float, etc.)
+        offset : int, optional
+            The offset in bytes. Usually corresponds to the number of bytes occupied by the header. 0 by default.
+        sampling_frequency : float, optional
+            The sampling frequency.
+        image_structure_to_axis : dict, optional
+            A dictionary indicating what axis corresponds to what in the memmap. The default values are:
+            dict(frame_axis=0, num_channels=1, rows=2, columns=3)
+            frame_axis=0 indicates that the first axis corresponds to the frames (usually time)
+            num_channels=1 here indicates that the first axis corresponds to the  n_channels.
+            rows=2 indicates that the rows in the image are in the second axis
+            columns=3 indicates that columns is the last axis or dimension in this structure.
+
+            Notice that this should correspond with frame_shape.
```

Comment:
This argument is an interesting proposition. But I wonder if there aren't better ways of combining this intention with the `frame_shape`, such as combininng the two in something like an `OrderedDictionary` where the order communicates the intended order of the axis?

Reply:
It could be. I thought about like an hour. Then decided that implementation as I wanted a working prototype. I am hoping something better comes to my head. My naive implementation in that direction got to a point where I needed the string version of the variable name and I always stop when I go there. 

Open to suggestions.

Reply:
How about a single argument that is a `Union[Tuple[int, int, int], Tuple[int, int, int]]` that is used to map the axes from the memmap to an roiextractors-enforced standard (which would make array shaping easier for things like `get_frames` and cross-extractor usage) of (frames, rows, columns, channels).

Also, I kind of like `channels` to be on the outer level here so reflect the common ordering of such axes for movie data (which is another extractor to be added soon).

Reply:
`get_frames` can be done in one line with the function `np.take` as long as we know which one is the frame_axis. 

That said, I am sympathetic to the idea to have a common inner standard in which the data should be forced for internal operations. 

Can you elaborate on how are you thinking though?
Let's say the example from the test:
```python
        num_frames = 50
        num_channels = 3
        rows = 10
        columns = 10
        memmap_shape = (num_channels, rows, columns, num_frames)
```

How are your tuples in this case?
Let's say that the reference is `(num_frames, num_channels, rows, columns)`.

Reply:
>  memmap_shape = (num_channels, rows, columns, num_frames)

The tuple I'm imaging, in this case, would be `(3, 1, 2, 0)` -> `(frames, rows, columns, channels)` since the rows/columns axes are already in the right places, but the channels/frames are reversed.

Another option to allow textual specification without the change for key errors that a standard dictionary would have for this kind of thing, could be either a `namedtuple` or `dataclass` (`dataclass` being the more modern one, I'd say). Check out both those standard (but more exotic) Python data types if you haven't seen them before.

Specifically, the `dataclass `(here, named 'AxisOrder' for example) approach would look, as an input, like `AxisOrder(frames=3, rows=1, columns=2, channels=0)`, which is close to the spirit of the `dict` version you proposed but has the critical difference that each key is forced to be one of those 4 choices (with, perhaps, `channels` being optional) whereas a normal `dict` can contain any string (well, actually, any hashable Python data type). This, and the `Enum` data type, are the approach we took for this type of stuff in the NWBInspector and it really does help prevent silly things like typos from going any further than they need to. Also they aren't mutable, so that's a plus over the `dict` approach, too.

Reply:
> > memmap_shape = (num_channels, rows, columns, num_frames)
> 
> The tuple I'm imaging, in this case, would be `(3, 1, 2, 0)` -> `(frames, rows, columns, channels)` since the rows/columns axes are already in the right places, but the channels/frames are reversed: 

So you have:
`Union[Tuple[int, int, int], Tuple[int, int, int]]`
There is a tuple on the left and one in the right.
Is the one in the left `(3, 1, 2, 0)` and the one in the right `(frames, rows, columns, channels)`?

> Another option to allow textual specification without the change for key errors that a standard dictionary would have for this kind of thing, could be either a `namedtuple` or `dataclass` (`dataclass` being the more modern one, I'd say). Check out both those standard (but more exotic) Python data types if you haven't seen them before.
> 
> Specifically, the `dataclass `(here, named 'AxisOrder' for example) approach would look, as an input, like `AxisOrder(frames=3, rows=1, columns=2, channels=0)`, which is close to the spirit of the `dict` version you proposed but has the critical difference that each key is forced to be one of those 4 choices (with, perhaps, `channels` being optional) whereas a normal `dict` can contain any string (well, actually, any hashable Python data type). This, and the `Enum` data type, are the approach we took for this type of stuff in the NWBInspector and it really does help prevent silly things like typos from going any further than they need to. Also they aren't mutable, so that's a plus over the `dict` approach, too.

Yes, I was explicitly thinking on using a data-class earlier today but my feeling is that it delegates too much definitions to the users. Tuples and dicts are easier to define and most people (specially scientist who are not well versed in OOD) understand them.  A data class simplifies our input and allows us to have checks on the spot but is a step further in complexity. I am on the fence about that trade-off.

The current solution is very simple as long as the format is on the "standard" (having the frames on the first axis) as it just demands structuring their data in a tuple. If their data is not in that format then and only then asks them to use a little bit more of complexity with a dict.

This is an interesting design problem.

Reply:
> but my feeling is that it delegates too much definitions to the users

If it helps, remember the user wouldn't define anything about the class - syntactically it would be just like specifying a dictionary, except it would immediately notify them if they forgot/failed to specify one of the keys, or if they mispelled a key. Basically, it's a dictionary with a hard-coded selection of keys they can associate values too.

Reply:
Yes, I was thinking that they will need to import it if we are talking about a data class. I am happy to prototype something though and see how that look / feels.